### ISSUE TITLE
Document output format for `.Rmarkdown` files

### ISSUE THREAD

**Issue by apreshill** [rstudio/blogdown#530]
(I also need to know the answer to this question!)

We have `blogdown::html_page` documented in the code, but it isn't clear how to knit an `.Rmarkdown` file to the appropriate `.markdown` file output format. I find myself trying to use the knit button to knit an `.Rmarkdown` post inside my own theme's exampleSite, and unable to knit to `.markdown` file I think because the project is not detected as a blogdown website by the IDE. The knit button knits straight to `.html`, and I cannot figure out how to ask for the `.markdown` output format.

**Comment by cderv** [comment]
I think this is related to another topic we discuss: 

* How could **blogdown** offer a full markdown mode ? 

By full markdown mode, the workflow where someone wants to use only the `.Rmarkdown` to `.markdown` mode of **blogdown** to let hugo handles all outputs to HTML. I guess in that case one would prefer to write `.Rmd` and output only `.md` file, without having to "play" with the extension. 

I think in other terms, this would be how to include **hugodown** behavior inside **blogdown** ? 

I am thinking more and more about that. That is not the place to develop maybe, but I think this is related because having a `goldmark_output()` or `hugo_output()` or else, would be very handy and coherent with how R Markdown works usually: I want to output another format, I use a different output format function. **blogdown** relies on extension for now for this... 

By the way, this is also thoughts I have because the IDE uses pandoc extensions to output [Goldmark format](https://github.com/rstudio/rstudio/blob/06503e51e2df2d183e055454eb8b51b8250a3ff6/src/gwt/panmirror/src/editor/src/api/pandoc_format.ts#L397) and **hugodown** [does to](https://github.com/r-lib/hugodown/blob/96e9700c141c27887c3801bb275796e6c65d9436/R/md-document.R#L143) (but differently it seems) - so I would think it is possible for **blogdown** to have a `md_document` variant to convert Pandoc Markdown to Goldmark markdown. 

Anyway, if this is interesting, we could discuss it more. 

Regarding your question now, from what I understand (it was a great opportunity to know more about **blogdown** internals, so sharing my understanding)

* The `.Rmarkdown` is only knitted to `.markdown` (no markdown conversion) as `run_pandoc = FALSE` in this case
https://github.com/rstudio/blogdown/blob/e1be8b9b98a1161e08580b48685686e0c38605c7/R/render.R#L194-L197


* But there are some post processing with `bookdown:::process_markdown` that I think assumes, the `.Rmarkdown` is written with Pandoc Markdown 
https://github.com/rstudio/blogdown/blob/e1be8b9b98a1161e08580b48685686e0c38605c7/R/render.R#L217
and another one if there are references: 
https://github.com/rstudio/blogdown/blob/e1be8b9b98a1161e08580b48685686e0c38605c7/R/render.R#L230-L231

Is this correct ? 
If so, this could be a base for a documentation on this internal behavior, with a version toward end users.

**Comment by apreshill** [comment]
This is indeed a "mode" that would benefit many users, as we mention in the [book](https://bookdown.org/yihui/blogdown/output-format.html#output-format):
> "The main advantage of using .Rmarkdown is that the output files are cleaner because they are Markdown files."

Additional benefits:

+ Populates the `{{ .TableOfContents }}` Hugo variable (https://gohugo.io/content-management/toc/)

+ Users can configure their TOC via the goldmark renderer (https://gohugo.io/getting-started/configuration-markup/#table-of-contents)

+ Users can benefit from Markdown render hooks via the goldmark renderer (https://gohugo.io/getting-started/configuration-markup/#markdown-render-hooks); like:

```
layouts
└── _default
    └── _markup
        ├── render-image.html
        ├── render-image.rss.xml
        └── render-link.html
```

It feels like `.md` should be one build method: https://bookdown.org/yihui/blogdown/methods.html#methods

`.Rmarkdown` --> `.markdown` is less portable and less familiar.

The `keep_md` YAML option was added, but this requires someone to add one or all html files to their `.gitignore` because of the Hugo trumping rules (`.html` > `.md`). I'm pretty sure this means that their local preview, though, will show the `.html` version, but the `.md` version will only show up after a Netlify build happens- you'd have to add html files to your Hugo build ignore *also*. The `.gitignore` approach is even harder if by chance you have a few actual `.html` content files on purpose, so you'd have to start ignoring individual files by name instead of using the wildcard `*.html`.

**Comment by maelle** [comment]
I would also like full Markdown support. A hack I've just tried
* Using a custom build method https://bookdown.org/yihui/blogdown/methods.html
* Having the build.R below to have index.Rmd knit to index.md with the hugodown output format

```r
path <- rstudioapi::getSourceEditorContext()$path
rmarkdown::render(
  path,
  output_format = hugodown::md_document()
)
file.remove(sub(".Rmd", ".html", path))
```

It would work with other Markdown formats, of course, but I tried the hack to get downlit highlighting.

**Comment by cderv** [comment]
I find it odd that you need to remove the html file - using `method = "custom"` should only run `build.R` and you only generate a md document. 🤔 It is as if another render is happening.

Also, for this type of use case, it would be interested to pass `build_rmd` as an argument to build.R script so that it can be used inside the script.
https://github.com/rstudio/blogdown/blob/28bb4a8a9f4aa0e7283368a82511d06a8428a702/R/render.R#L62-L67
this would be a way to get the path of the file without having to rely on **rstudioapi** (which would currently not work if you try to use `build_site()` at command line and the Rmd needs to be knitted).

**Comment by maelle** [comment]
I thought the HTML file came from hugodown but I'm not so sure.

I actually use the knit button. (which might explain the HTML creation?)

**Comment by yihui** [comment]
> I find myself trying to use the knit button to knit an `.Rmarkdown` post inside my own theme's exampleSite, and unable to knit to `.markdown` file I think because the project is not detected as a blogdown website by the IDE.

@apreshill If you are developing a theme, it's a little tricky to make the RStudio IDE recognize the type of the project and render posts under `exampleSite/content/`. There are two things that you need to do:

1. `setwd('exampleSite')`: https://github.com/yihui/hugo-prose/blob/master/.Rprofile
2. change the default root dir to `WebsitePath: exampleSite`: https://github.com/yihui/hugo-prose/blob/master/hugo-prose.Rproj

This should solve your original problem. The rest of replies brought up another issue, which I just tried to address:

> It feels like `.md` should be one build method: https://bookdown.org/yihui/blogdown/methods.html#methods

@apreshill That makes sense. I left the option `blogdown.method` open in the original design of `build_site()` (with a vague feeling that I'd need to support more methods in the future), and now I'm glad that I did. I just provided `options(blogdown.method = "markdown")`. Once you set this option, `.Rmd` will be rendered to `.md` (under the hood, `.Rmd` receives the same treatment as `.Rmarkdown`, which was explained in detail by @cderv above).

I guess that will also solve @maelle's problem (no need to hack with `method = "custom"`; however, at this point, you still don't have **downlit** support with `method = "markdown"`).

**Comment by maelle** [comment]
I can't get ".Rmd will be rendered to .md" to work.

* I re-installed blogdown, removed my build.R lines and changed the method in the Rprofile.
* I re-started R and even RStudio.

The Rmd still gets knit to html :thinking: https://github.com/maelle/apresacademic/tree/main/content/blog/2021-01-05-hi-blogdown

Sorry if I'm getting something wrong. I'm very excited about this update!

**Comment by cderv** [comment]
It seems there is something off in RStudio IDE when using the knit button. 

Using `blogdown::build_site(build_rmd = TRUE)` will correctly render .Rmd to .md when the option is set. 
Same with `rmarkdown::render_site('<absolute_path_to_your_Rmd>')` in the R console. 
But Knitting with the button will output a HTML. 

This happens only with .Rmd to .md - .Rmarkdown are correctly rendered to .markdown. 

Still looking into this 🤔

**Comment by cderv** [comment]
Continuing the discussion in its own issue https://github.com/rstudio/blogdown/issues/562

This happens with `method != "html"` and not only with this new feature (https://github.com/rstudio/blogdown/commit/b84bc8f538fe00861fa62e0c54c3468126a241e0) - I think it was also an issue with `method = "custom"` -> Knit button does not fully work with **blogdown** yet.

**Comment by yihui** [comment]
Sounds like a bug. Sorry I didn't test it yesterday since it was too late at night. I'm investigating now.

**Comment by cderv** [comment]
Just to be clear, I stumbled upon this bug before you added the feature here, while testing what @maelle was trying (https://github.com/rstudio/blogdown/issues/530#issuecomment-754483753). 

As it was happening also with the new method, it confirmed something was off. 

I investigated this morning but was missing some information as I am not sure what the Knit button is really doing in RStudio IDE. 

Happy to discuss it further !
